### PR TITLE
e2e: adding e2e upgrade test for ibc-go/v6

### PR DESCRIPTION
**Pull request by damiannolan**
## Description

- Adding E2E upgrade test for `ibc-go/v6`

### Workflow

- https://github.com/cosmos/ibc-go/actions/runs/3234968284

### Local test setup for work-in-progress

1. Pull `ibc-go-icad` docker images

```
docker pull ghcr.io/cosmos/ibc-go-icad:v0.3.5
docker pull ghcr.io/cosmos/ibc-go-icad:v0.4.0
```

2. Export `e2e` env variables

```
export CHAIN_BINARY=icad
export CHAIN_IMAGE=ghcr.io/cosmos/ibc-go-icad
export CHAIN_A_TAG=v0.3.5
```

3. Run `e2e` upgrade test

```
cd e2e
make e2e-test entrypoint=TestUpgradeTestSuite test=TestV5ToV6ChainUpgrade
```

closes: #2188 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: e2e/tests/upgrades/v6/upgrade_test.go [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+	s.Require().GreaterOrEqual(height, haltHeight+blocksAfterUpgrade, "height did not increment enough after upgrade")
+}
+
+// TODO: Swap e2e transfer flow logic for interchain accounts
+func (s *UpgradeV6TestSuite) TestV5ToV6ChainUpgrade() {
+	t := s.T()
+
+	oldVersion := "pr-140"
+	targetVersion := "pr-139"
```
Comment (damiannolan) [inline e2e/tests/upgrades/v6/upgrade_test.go:80]: See: 
- https://github.com/cosmos/interchain-accounts-demo/pull/140
- https://github.com/cosmos/interchain-accounts-demo/pull/139

File: e2e/tests/upgrades/v6/upgrade_test.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+		s.Require().NoError(err)
+
+		expected := testvalues.IBCTransferAmount
+		s.Require().Equal(expected, actualBalance)
+	})
+
+	s.Require().NoError(test.WaitForBlocks(ctx, 5, chainA, chainB), "failed to wait for blocks")
+
+	t.Run("upgrade chainA", func(t *testing.T) {
+		s.UpgradeChain(ctx, chainA, chainAUpgradeProposalWallet, "ibc-go/v6", targetVersion)
```
Comment (damiannolan) [inline e2e/tests/upgrades/v6/upgrade_test.go:139]: Will update the `ibc-go/v6` magic string with the `UpgradeName` const

File: e2e/tests/upgrades/v6/upgrade_test.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+	err = test.WaitForBlocks(timeoutCtx, int(blocksAfterUpgrade), chain)
+	s.Require().NoError(err, "chain did not produce blocks after upgrade")
+
+	height, err = chain.Height(ctx)
+	s.Require().NoError(err, "error fetching height after upgrade")
+
+	s.Require().GreaterOrEqual(height, haltHeight+blocksAfterUpgrade, "height did not increment enough after upgrade")
+}
+
+// TODO: Swap e2e transfer flow logic for interchain accounts
```
Comment (damiannolan) [inline e2e/tests/upgrades/v6/upgrade_test.go:75]: Refactor the transfer logic to follow ICS27 workflow.

- [ ] Create an ICS27 channel using an existing auth module. E.g. intertx
- [ ] Verify SendTx behaves as expected
- [x] Create software upgrade proposal
- [x] Execute proposal and upgrade to new chain binary
- [x] Migrations are run via upgrade handler and automatic migrations via consensus version increment.
- [ ] Verify SendTx behaves as expected

File: e2e/tests/upgrades/v6/upgrade_test.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+	suite.Run(t, new(UpgradeV6TestSuite))
+}
+
+type UpgradeV6TestSuite struct {
+	testsuite.E2ETestSuite
+}
+
+// UpgradeChain upgrades a chain to a specific version using the planName provided.
+// The software upgrade proposal is broadcast by the provided wallet.
+func (s *UpgradeV6TestSuite) UpgradeChain(ctx context.Context, chain *cosmos.CosmosChain, wallet *ibc.Wallet, planName, upgradeVersion string) {
```
Comment (damiannolan) [inline e2e/tests/upgrades/v6/upgrade_test.go:36]: I think we can move this out and maintain at the top level `upgrades` directory

File: e2e/tests/upgrades/v6/upgrade_test.go [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+	"github.com/cosmos/ibc-go/e2e/testsuite"
+	"github.com/cosmos/ibc-go/e2e/testvalues"
+)
+
+const (
+	haltHeight         = uint64(100)
+	blocksAfterUpgrade = uint64(10)
+)
+
+func TestUpgradeV6TestSuite(t *testing.T) {
```
Comment (chatton) [inline e2e/tests/upgrades/v6/upgrade_test.go:26]: there are some assumptions being made now that all upgrade tests live under the `TestUpgradeTestSuite` function. See [this workflow](https://github.com/cosmos/ibc-go/blob/5f16e07990f124b728f51c2beb1a8aeb1a7ff1a1/.github/workflows/e2e-upgrade.yaml#L23).

There's nothing stopping us from changing this, just highlighting this as the upgrade tests won't automatically start running this with the current structure.

Comment (chatton) [reply e2e/tests/upgrades/v6/upgrade_test.go:26]: On a related note, currently, we **don't** run upgrade tests on PRs, this is achieved by explicitly excluded them. We do that

[here for forks](https://github.com/cosmos/ibc-go/blob/5f16e07990f124b728f51c2beb1a8aeb1a7ff1a1/.github/workflows/e2e-fork.yml#L27) and [here for regular prs](https://github.com/cosmos/ibc-go/blob/036c09e632ea92b39e25c562037d30bca316a64e/.github/workflows/e2e.yaml#L51)

File: .github/workflows/e2e-manual-icad.yaml [hunk 6]
```diff
@@ -27,6 +28,8 @@ on:
           - v0.3.4
           - v0.2.3
           - v0.1.4
+          - pr-139
+          - pr-140
```
Comment (chatton) [inline .github/workflows/e2e-manual-icad.yaml:32]: we'll need to come up with a better way of doing this when we need new versions 🤔

File: e2e/tests/upgrades/upgrade_test.go [hunk 7]
```diff
@@ -7,15 +7,23 @@ import (
 	"testing"
 	"time"
 
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
 	upgradetypes "github.com/cosmos/cosmos-sdk/x/upgrade/types"
+	v6Upgrades "github.com/cosmos/interchain-accounts/app/upgrades/v6"
```
Comment (chatton) [inline e2e/tests/upgrades/upgrade_test.go:13]: [nit] should this be `v6upgrades`? (all lower case)

File: .github/workflows/e2e-manual-icad.yaml [hunk 8]
```diff
@@ -12,6 +12,7 @@ on:
         options:
           - TestInterchainAccountsTestSuite
           - TestIncentivizedInterchainAccountsTestSuite
+          - TestUpgradeV6TestSuite
```
Comment (chatton) [inline .github/workflows/e2e-manual-icad.yaml:15]: I believe this should be `TestUpgradeTestSuit` now that both upgrade tests are on the same test suite.

File: .github/workflows/e2e.yaml [hunk 9]
```diff
@@ -48,4 +48,4 @@ jobs:
         relayer-tag: "v2.0.0"
         chain-binary: "simd"
         # on regular PRs we won't run interchain account or upgrade tests.
-        test-exclusions: "TestInterchainAccountsTestSuite,TestIncentivizedInterchainAccountsTestSuite,TestUpgradeTestSuite"
+        test-exclusions: "TestInterchainAccountsTestSuite,TestIncentivizedInterchainAccountsTestSuite,TestUpgradeTestSuite,TestUpgradeV6TestSuite"
```
Comment (chatton) [inline .github/workflows/e2e.yaml:51]: this additional exclusion isn't required anymore

**Review by chatton — COMMENTED** [review]
Few small changes and then I think we should be good to merge! great work on this one ❤️

File: .github/workflows/e2e-manual-icad.yaml (hunk 8 above)
Comment (damiannolan) [reply .github/workflows/e2e-manual-icad.yaml]: Done

File: .github/workflows/e2e.yaml (hunk 9 above)
Comment (damiannolan) [reply .github/workflows/e2e.yaml]: Nice catch! :D

**Review by colin-axner — APPROVED** [review]
wahoo!

**Review by chatton — APPROVED** [review]
LGTM nice work 🟢

**Pull request merged**